src/unwinder/i386/unwinder_i386.cpp

```diff
@@ -17,11 +17,6 @@ BOOL OOPStackUnwinderX86::Unwind(T_CONTEXT* pContextRecord, T_KNONVOLATILE_CONTE
     rd.SP = pContextRecord->Esp;
     rd.PCTAddr = (UINT_PTR)&(pContextRecord->Eip);
 
-    if (pContextPointers)
```

Comment:
This change doesn't look like a correct fix. You are essentially throwing away the previous state of the context pointers. There might be cases when a context pointer survives several unwinds. For example:
* There is a call chain of functions A->B->C->D.
* Only D saves edi, since it uses it, none of A, B and C does.
* We start unwinding at D. 
* First unwind unwinds to C and gives us a correct edi context pointer
* Second unwind unwinds to B, but we've lost the edi context pointer as the unwind ignored the context pointers passed to it.

I am not sure why UnwindEbpDoubleAlignFrame actually gets the Ebp by dereferencing the GetEbpLocation() instead of getting it from the pContext->pCurrentContext->Ebp. Do you remember that? And if not, can you please try to check if it works and fixes the problem? I am not currently setup for debugging x86 Linux, so it would take me some time to try it myself.

Reply:
Yes, thank you! I missed that pContextPointers is inout parameter. I don’t remember why pointers is used for ebp, it is necessary to investigate. I will check you change, thank you!

Reply:
With the following change crash doesn't occur:
```diff
diff --git a/src/vm/eetwain.cpp b/src/vm/eetwain.cpp
index 6bf53738fd..4d66c468d5 100644
--- a/src/vm/eetwain.cpp
+++ b/src/vm/eetwain.cpp
@@ -3873,8 +3873,8 @@ bool UnwindEbpDoubleAlignFrame(
 
     _ASSERTE(info->ebpFrame || info->doubleAlign);
 
-    const unsigned curESP =  pContext->SP;
-    const unsigned curEBP = *pContext->GetEbpLocation();
+    const unsigned curESP = pContext->SP;
+    const unsigned curEBP = pContext->pCurrentContext->Ebp;
```

Reply:
Also I looked at using `ContextPointers` in amd64 implementation of `RtlVirtualUnwind` and it is used only as out parameter:
```cpp
if (ARGUMENT_PRESENT(ContextPointers)) {
    ContextPointers->IntegerContext[OpInfo] = IntegerAddress;
}
```

Reply:
Well, it is still in-out on amd64 - it doesn't destroy the previous state.

Reply:
yes, in this meaning. What do you think about updated patch?